Consumer wasm api

nicholastmosher opened the pull request:
Closes https://github.com/infinyon/fluvio-client-wasm/issues/32

- Adds `SmartStreamModule` type for WASM binaries given to the consumer by users
- Turn `ConsumerConfig` into a proper builder that can take WASM by binary or base64 (for web use-cases)

sehz commented:
CI failing?

sehz commented:
Can we not require bump version? Can deprecated existing API

simlay commented on src/client/src/consumer.rs:
```diff
@@ -410,12 +416,13 @@ impl PartitionConsumer {
-                stream_request.wasm_module = config.wasm_module;
-            } else {
+        if let Some(wasm) = config.smartstream_filter {
+            println!("USING SMARTSTREAM WITH WASM SIZE: {}", wasm.binary.len());
```
I think you want `debug!`.

nicholastmosher replied:
Ah, I actually need to remove that, good catch

nicholastmosher commented:
@sehz I reverted deleting the old methods and I marked them as deprecated. Changed crate bump as `0.8.5 -> 0.8.6`. I also think I fixed CI, but we'll see. CHANGELOG has also been updated. If CI passes this should be good to go.

sehz commented on src/client/src/consumer.rs:
```diff
@@ -564,38 +571,159 @@ static MAX_FETCH_BYTES: Lazy<i32> = Lazy::new(|| {
 });
 
+#[derive(Debug, Clone)]
+pub struct SmartStreamModule {
```
Instead of creating another wrapper, rather use existing fluvio data type `RecordData` instead of bytes.  This will also come in handy when storing WASM bytes in the Fluvio later on.

nicholastmosher replied:
I think it would not be a good idea to overload `RecordData` to mean two different things. I also think that it is actually a good idea to have this `SmartStreamModule` type because we can expand on it in the future, e.g. validate that it is correct WASM before instantiating it. If you would prefer not to have a new type for this right now, I can turn it back into a `Vec<u8>`, but I think using `RecordData` is a bad idea.

sehz commented on src/client/src/consumer.rs:
```diff
@@ -564,38 +571,159 @@ static MAX_FETCH_BYTES: Lazy<i32> = Lazy::new(|| {
+    }
+
+    /// Instantiate a SmartStream from a base64-encoded string
+    pub fn from_base64<S: Into<String>>(string: S) -> Result<Self, FluvioError> {
```
This seems too premature to add as supporting API.  This conversion can be done by caller.

sehz commented on src/client/src/consumer.rs:
```diff
@@ -564,38 +571,159 @@ static MAX_FETCH_BYTES: Lazy<i32> = Lazy::new(|| {
 
     /// set wasm filter
+    #[deprecated(since = "0.8.6", note = "Use 'ConsumerConfig::builder()' instead")]
     pub fn with_wasm_filter(mut self, bytes: Vec<u8>) -> Self {
```
use RecordData

nicholastmosher replied:
See my comment above

sehz commented:
https://github.com/infinyon/fluvio-client-wasm/issues/32 need further work.   We need further design work on how to configure and pass WebAssembly.  So we should close this for now

nicholastmosher commented:
@sehz I minimized this some more based on discussion today, let me know if it looks alright.

sehz commented:
CI failing?

sehz reviewed (changes requested):
Given this seems only slight better API but may add more tech debt, is this PR necessary? I would rather tackle this when we need to start adding compression or other features.

SmartStream config may need it's own builder (SmartStreamConfigBuilder?)

nicholastmosher commented on src/client/src/consumer.rs:
```diff
@@ -565,33 +572,67 @@ static MAX_FETCH_BYTES: Lazy<i32> = Lazy::new(|| {
-    wasm_module: Vec<u8>,
+    /// A WASM module to use as a SmartStream filter
+    #[builder(setter(into, strip_option), default)]
+    smartstream_filter: Option<Vec<u8>>,
```
Option makes more sense here IMO. Seeing this in the struct as `smartstream_filter: Vec<u8>` makes me think that the field will always be populated, whereas an `Option<Vec<u8>>` semantically lets me know that I should expect to check whether a value is present or not. Note that this also does not have any runtime cost, because any type containing a pointer (like vec does) inside an Option will be niche-optimized: https://stackoverflow.com/questions/16504643/what-is-the-overhead-of-rusts-option-type

nicholastmosher commented:
I noticed that now that I stopped using the `Rc<RefCell<...>>` pattern in fluvio-client-wasm that these changes aren't necessary any more. We can revisit config in the future.
